File: сontest 5/B.cs
using System;
using System.Text;
    
internal class B
{
    static void Main(string[] args)
    {
        var t = Console.ReadLine();
        var q = int.Parse(Console.ReadLine());
        for (int i = 0; i < q; i++)
        {
            var s = Console.ReadLine();
            var (cnt, result) = zFunction(s, t);
            if (cnt == 0) Console.WriteLine(cnt);
            else Console.WriteLine($"{cnt} {result}");
        }
    }

    private static (int, string) zFunction(string s, string t)
    {
        var cnt = 0;
        var str = $"{s}${t}";
        var result = new StringBuilder();
        var zf = new int[str.Length];
        int l = 0, r = 0;
        for (int i = 1; i < str.Length; i++)
        {
            zf[i] = Math.Max(0, Math.Min(r - i, zf[i - l]));
            while (i + zf[i] < str.Length && str[zf[i]] == str[zf[i] + i])
            {
                zf[i]++;
            }

            if (zf[i] + i > r)
            {
                l = i;
                r = i + zf[i];
            }

            if (zf[i] == s.Length)
            {
                result.Append((i - s.Length - 1).ToString() + " ");
                cnt++;
            }
        }

        return (cnt, result.ToString());
    }
}

File: сontest 5/B.py
# WARNING!!!
# ДАННАЯ ЗАДАЧА НЕ ПРОХОДИТ НА PYTHON С ОПТИМАЛЬНЫМ РЕШЕНИЕМ ЧЕРЕЗ Z-ФУНКЦИЮ
# АНАЛОГИЧНЫЙ АЛГОРИТМ ЗАХОДИТ НА C#

def z_function(s, t):
    result = ''
    cnt = 0
    str = f'{s}${t}'
    zf = len(str) * [0]
    l, r = 0, 0

    for i in range(1, len(str)):
        zf[i] = max(0, min(r - i, zf[i - l]))
        while i + zf[i] < len(str) and str[zf[i]] == str[i + zf[i]]:
            zf[i] += 1
        if i + zf[i] > r:
            l = i
            r = i + zf[i]
        if zf[i] == len(s):
            result += f'{i - len(s) - 1} '
            cnt += 1

    return cnt, result.rstrip()


t = input()
q = int(input())

for _ in range(q):
    s = input().strip()
    cnt, result = z_function(s, t)

    print(cnt) if cnt == 0 else print(f'{cnt} {result}')